facebookresearch/ParlAI
# [scale] Implement activation checkpointing for transformers.

**Patch description**
Leverage Fairscale to add in [activation checkpointing](https://fairscale.readthedocs.io/en/stable/api/nn/checkpoint/checkpoint_activations.html). Activation checkpointing works by throwing away the activations during the forward pass, and recomputing them just-in-time during the backward pass, thereby trading away compute for memory savings.

It is particularly powerful when combined with FSDP. Utilized with `--checkpoint-activations true`

**Testing steps**
New CI test to just ensure that the path doesn't crash. Due to the simplicity of the call, I rely on fairscale's tests for correctness.

Manual testing: `parlai train -m bart -t taskmaster2 -bs 4 -dynb full -tstep 100 --checkpoint-activations true`. Took some very non-scientific measurements but were exactly as expected:

| Checkpoint | Memory (nvidia-smi) | Updates/sec |
|-------------|-------|----|
| No (baseline) | 10823mb | 1.01 u/s |
| Yes | 7853mb |  .77 u/s |
